marker_api/src/ast/item.rs

```diff
@@ -238,42 +240,182 @@ use impl_item_data;
+/// // An item with a restricted scope
+/// pub(crate) fn star() {}
+///
+/// pub trait Planet {
+///     // An item without a visibility. But it still has the semantic visibility
+///     // of `pub` as this is inside a trait declaration.
+///     fn mass();
+/// }
+/// ```
 #[repr(C)]
+#[derive(Debug)]
+#[cfg_attr(feature = "driver-api", derive(typed_builder::TypedBuilder))]
 pub struct Visibility<'ast> {
+    #[cfg_attr(feature = "driver-api", builder(setter(skip), default))]
     _lifetime: PhantomData<&'ast ()>,
-    _item_id: ItemId,
+    #[cfg_attr(feature = "driver-api", builder(setter(into)))]
+    #[cfg_attr(feature = "driver-api", builder(default = FfiOption::None))]
+    span: FfiOption<SpanId>,
+    kind: VisibilityKind,
+    // * The rustc model makes sense, so have a kind with a restriction
+    // * have according methods, that check for `pub(crate)`
+    // * document possible equivalency between `pub(crate)` and `pub(super)`
+    // * Document that things can be reexported
+    // * Test how visibility only to lower scopes works `pub(self::sub_module)`
 }
 
-impl<'ast> Debug for Visibility<'ast> {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        f.debug_struct("Visibility {{ /* WIP: See rust-marker/marker#26 */}}")
-            .finish()
+impl<'ast> Visibility<'ast> {
+    pub fn span(&self) -> Option<&Span<'ast>> {
+        self.span.copy().map(|span| with_cx(self, |cx| cx.span(span)))
     }
-}
 
-#[cfg(feature = "driver-api")]
-impl<'ast> Visibility<'ast> {
-    pub fn new(item_id: ItemId) -> Self {
-        Self {
-            _lifetime: PhantomData,
-            _item_id: item_id,
+    /// Returns `true` if the item is declared as public, without any restrictions.
+    ///
+    /// ```
+    /// // This returns `true`
+    /// pub fn unicorn() {}
+    ///
+    /// // This returns `false`, since the visibility is restricted to a specified path.
+    /// pub(crate) fn giraffe() {}
+    ///
+    /// // This returns `false`, since the visibility is not defined
+    /// fn dragon() {}
+    /// ```
+    ///
+    /// See [`Visibility::is_pub_with_path`] to detect pub declarations with a
+    /// defined path.
+    pub fn is_pub(&self) -> bool {
+        matches!(self.kind, VisibilityKind::Public | VisibilityKind::DefaultPub)
+    }
+
+    /// Returns `true` if the item is declared as `pub(...)` with a path in brackets
+    /// that defines the scope, where the item is visible.
+    pub fn is_pub_with_path(&self) -> bool {
+        matches!(self.kind, VisibilityKind::Path(_) | VisibilityKind::Crate(_))
+    }
+
+    /// Returns `true` if the visibility is declared as `pub(crate)`. This is a
+    /// special case of the `pub(<path>)` visibility.
+    ///
+    /// This function checks if the visibility is restricted and the defined path
+    /// belongs to the root module of the crate. Meaning, that this can also be `true`,
+    /// if the visibility uses `pub(super)` to travel up to the crate root.
```

Comment:
So this will return true if the path refers to the root of the crate. That looks more like the semantic property of this type, not the ast. Maybe it makes sense to make the ast/sem distinction more exlicit? Is it even possible to check if the visibility is `pub(crate)` syntactically (and not `pub(super)` or `pub(in ...)` that lead to the root of the crate).

Reply:
Rustc sadly erases the information how the path has been specified. That's why this type provides semantic information when it comes to this. I could rearrange it, so that we have a semantic visibility, and the AST visibility, basically just wraps that one and optionally provides a `Span`. What do you think about this structure?

Reply:
I see, that's a pity. Anyway, since we don't have a different from the original AST representation of the visibility, I think it does make sense